### PR TITLE
F/virtual machine extension by ID fixes #2456

### LINKED ISSUE #2456: azurerm_virtual_machine_extension should also accept virtual_machine_id instead of virtual_machine_name
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Currently the azure_virtual_machine_extension resource requires to specify the concerned VM with the parameter "virtual_machine_name". This is not consistent with other resources that require the "virtual_machine_id" parameter. Also, I think that it would be safer to use the id instead of the name.

When developing a module that has both a disk attachment and a vm extension, this is a bit cumbersome because it requires my module interface to require both the id and the name of the VM. One possible solution would also be to create a azure_virtual_machine_extension data source to resolve the name from the id (but still: having only ids would be a cleaner interface in my opinion).

### New or Affected Resource(s)

* azure_virtual_machine_extension

### PR DESCRIPTION
**Pull request by jackofallops**
This change updates the VM extension resource to use the `virtual_machine_id` and deprecates the use of `virtual_machine_name`.

### CODE REVIEW

File: azurerm/internal/services/compute/registration.go [hunk 1]
```diff
@@ -45,7 +45,7 @@ func (r Registration) SupportedResources() map[string]*schema.Resource {
 		"azurerm_shared_image":                         resourceArmSharedImage(),
 		"azurerm_snapshot":                             resourceArmSnapshot(),
 		"azurerm_virtual_machine_data_disk_attachment": resourceArmVirtualMachineDataDiskAttachment(),
-		"azurerm_virtual_machine_extension":            resourceArmVirtualMachineExtensions(),
+		"azurerm_virtual_machine_extension":            resourceArmVirtualMachineExtension(),
```
Comment (katbyte) [inline azurerm/internal/services/compute/registration.go]: Nice catch 👍

File: azurerm/internal/services/compute/resource_arm_virtual_machine_extension.go [hunk 2]
```diff
@@ -85,26 +90,67 @@ func resourceArmVirtualMachineExtensions() *schema.Resource {
...
+
+		virtualMachineName = v.Name
+		resourceGroup = v.ResourceGroup
+
+		virtualMachine, err := vmClient.Get(ctx, resourceGroup, virtualMachineName, "")
+		if err != nil {
+			return fmt.Errorf("Error getting Virtual Machine %q (Resource Group %q): %+v", name, resourceGroup, err)
+		}
+
+		location = *virtualMachine.Location
```
Comment (katbyte) [inline azurerm/internal/services/compute/resource_arm_virtual_machine_extension.go]: I think we'll need to nil check this

File: azurerm/internal/services/compute/tests/resource_arm_virtual_machine_extension_test.go [hunk 3]
```diff
@@ -246,6 +274,111 @@ resource "azurerm_virtual_machine" "test" {
...
+
+  tags = {
+    environment = "Production"
+  }
+}
+`, data.RandomInteger, data.Locations.Primary, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger)
+}
+
+// TODO: Remove in 2.0
+func testAccAzureRMVirtualMachineExtension_deprecated(data acceptance.TestData) string {
```
Comment (katbyte) [inline azurerm/internal/services/compute/tests/resource_arm_virtual_machine_extension_test.go]: THis seems to be testing the new property so shouldn't be removed in 2.0/deprecated? 🤔

File: azurerm/internal/services/compute/tests/resource_arm_virtual_machine_extension_test.go [hunk 4]
```diff
@@ -40,6 +40,34 @@ func TestAccAzureRMVirtualMachineExtension_basic(t *testing.T) {
...
+				),
+			},
+			data.ImportStep("protected_settings"),
+			{
+				Config: testAccAzureRMVirtualMachineExtension_deprecatedUpdate(data),
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMVirtualMachineExtensionExists(data.ResourceName),
+					resource.TestMatchResourceAttr(data.ResourceName, "settings", regexp.MustCompile("whoami")),
+				),
+			},
```
Comment (katbyte) [inline azurerm/internal/services/compute/tests/resource_arm_virtual_machine_extension_test.go]: And another import step here

File: azurerm/internal/services/compute/resource_arm_virtual_machine_extension.go [hunk 5]
```diff
@@ -40,16 +40,21 @@ func resourceArmVirtualMachineExtensions() *schema.Resource {
 				Required: true,
 				ForceNew: true,
 			},
-
```
Comment (katbyte) [inline azurerm/internal/services/compute/resource_arm_virtual_machine_extension.go]: minor, generally we leave spaces between properties now, but not a blocker.

File: azurerm/internal/services/compute/tests/resource_arm_virtual_machine_extension_test.go (hunk 3 above)
Comment (jackofallops) [reply azurerm/internal/services/compute/tests/resource_arm_virtual_machine_extension_test.go]: This config uses the deprecated fields (name / location) which won't be available post-2.0, now commented as such.

**Pull request merged**